### PR TITLE
CMake modernization

### LINKED ISSUE #109: Using "modern" CMake in Ginkgo
I always had trouble finding good and up-to-date tutorials about CMake, and how to properly do stuff there (especially the things  that seem to be possible to do in multiple ways).

I think I finally managed to find a good collection of resources [here](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1).
Just by taking a quick look I already figured out that we're doing some of the things the wrong way which will prevent us in playing nicely with other software.

Someone should go over the materials in the link, and update our build system to follow the rules.

### PR DESCRIPTION
**Pull request by tcojean**
Modernize and improve Ginkgo's CMake settings. This is part of a series of changes to improve our CMake installation and ensure full compatibility with xSDK.

+ Slightly reorganize the main CMakeLists.txt,
+ Ensure 4 space alignment are used everywhere,
+ Prefer the specific `Ginkgo_SOURCE_DIR` and related variables as much as possible,
+ Use `target_include_directories` instead of the global `include_directories`,
+ Use `target_compile_features` instead of the global settings,
+ Move the header installation to the `ginkgo_install` function,
+ Use `target_sources` instead of the source variable,
+ Use `target_compile_options` instead of tweaking the `CMAKE_<LANG>_FLAGS`,
+ Use `target_compile_definitions` instead of the global `add_definitions`.

Closes #109.

Requirements:
+ [x] Merge #214
+ [x] Merge #216 
+ [x] Rebase, check new changes and retest this MR

This MR is on top of #216. To get a clean view of the changes, look at the latest commit only.

### CODE REVIEW

**Comment by tcojean** [comment]
This can now be reviewed.

File: CMakeLists.txt [hunk 1]
```diff
@@ -23,28 +31,22 @@ set(GINKGO_CUDA_ARCHITECTURES "Auto" CACHE STRING
...
 
-set(CMAKE_CXX_STANDARD 11)
-set(CMAKE_CXX_STANDARD_REQUIRED ON)
-
 
 # Needed because of a known issue with CUDA while linking statically.
 # For details, see https://gitlab.kitware.com/cmake/cmake/issues/18614
 if((NOT BUILD_SHARED_LIBS) AND GINKGO_BUILD_CUDA)
     enable_language(CUDA)
 endif()
```
Comment (thoasm) [inline CMakeLists.txt]: Interestingly, I don't need this `if` statement (including the `enable_language(CUDA)`) anymore on both my Arch Linux (cmake version 3.13.4) and on the CI system (cmake version 3.13.1) when compiling with static libraries.
Maybe some of your changes also affected this.

Do you have access to older CMake versions to try it? If it is still not needed, I think we can drop this.

File: install_test/CMakeLists.txt [hunk 2]
```diff
@@ -4,13 +4,11 @@ project(InstallTest LANGUAGES CXX)
 
 # `enable_language(CUDA)` is needed when ginkgo was statically build with CUDA
```
Comment (thoasm) [inline install_test/CMakeLists.txt]: Same here, I don't need this anymore on both CI and my Arch Linux when dealing with static libraries. I think we should be able to drop this in combination with the note in the [`README.md`](https://github.com/ginkgo-project/ginkgo/blob/develop/README.md#installing-ginkgo)

**Comment by thoasm** [comment]
On the CI system in a docker container (CMake version 3.11.1), I get the exact same result. I guess we can drop these lines in both the main [`CMakeLists.txt`](https://github.com/ginkgo-project/ginkgo/blob/cmake_modernization/CMakeLists.txt#L44) and comment in the [`install_test/CMakeList.txt`](https://github.com/ginkgo-project/ginkgo/blob/cmake_modernization/install_test/CMakeLists.txt#L5)

**Pull request merged**